test/functional/single_pages_controller_test.rb

```diff
@@ -181,11 +98,238 @@ def setup
+      assert response_body.key?('uuid'), msg = "Response body is expected to have a 'uuid' key"
+      cache_uuid = response_body['uuid']
 
       get :download_samples_excel, params: { uuid: cache_uuid }
       assert_response :ok, msg = 'Unable to generate the excel'
     end
   end
+
+  test 'invalid file extension should raise exception' do
+    with_config_value(:project_single_page_enabled, true) do
+      file_path = Rails.root.join('test/fixtures/files/upload_single_page/00_wrong_format_spreadsheet.ods')
+      file = fixture_file_upload(file_path, 'application/vnd.oasis.opendocument.spreadsheet')
+
+      project, source_sample_type = setup_file_upload.values_at(
+        :project, :source_sample_type
+      )
+
+      post :upload_samples, params: { file: file, project_id: project.id,
+                                      sample_type_id: source_sample_type.id }
+
+      assert_response :bad_request
+      assert_equal flash[:error], "Please upload a valid spreadsheet file with extension '.xlsx'"
+    end
+  end
+
+  test 'Should prevent to upload to the wrong Sample Type' do
+    with_config_value(:project_single_page_enabled, true) do
+      project, sample_collection_sample_type = setup_file_upload.values_at(
+        :project, :sample_collection_sample_type
+      )
+
+      file_path = Rails.root.join('test/fixtures/files/upload_single_page/01_combo_update_sources_spreadsheet.xlsx')
+      file = fixture_file_upload(file_path, 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet')
+
+      post :upload_samples, as: :json, params: { file:, project_id: project.id,
+                                                 sample_type_id: sample_collection_sample_type.id }
+
+      assert_response :bad_request
+    end
+  end
+
+  test 'Should not process invalid workbooks' do
+    with_config_value(:project_single_page_enabled, true) do
+      project, source_sample_type = setup_file_upload.values_at(
+        :project, :source_sample_type
+      )
+
+      file_path = Rails.root.join('test/fixtures/files/upload_single_page/02_invalid_workbook.xlsx')
+      file = fixture_file_upload(file_path, 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet')
+
+      post :upload_samples, as: :json, params: { file:, project_id: project.id,
+                                                 sample_type_id: source_sample_type.id }
+
+      assert_response :bad_request
+    end
+  end
+
+  test 'Should update, create and detect duplicate sources when uploading to a source Sample Type' do
+    with_config_value(:project_single_page_enabled, true) do
+      project, source_sample_type = setup_file_upload.values_at(
+        :project, :source_sample_type
+      )
+
+      file_path = Rails.root.join('test/fixtures/files/upload_single_page/01_combo_update_sources_spreadsheet.xlsx')
+      file = fixture_file_upload(file_path, 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet')
```

Comment:
for the file_path you just need to pass the path relative to the `test/fixtures/files`, and it will do the join for you. So you just need to pass `upload_single_page/01_combo_update_sources_spreadsheet.xlsx`